File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/count_nz_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class CountNZCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            CountNZCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.SZ,
            Condition.SNZ,
            Condition.SPL,
            Condition.SMI,
            Condition.MAX,
            Condition.NMAX,
            Condition.TRUE,
        }

File: golang/uPIMulator/src/linker/kernel/instruction/cc/count_nz_cc.go
package cc

import (
	"errors"
)

type CountNzCc struct {
	condition Condition
}

func (this *CountNzCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		Z:    true,
		NZ:   true,
		XZ:   true,
		XNZ:  true,
		SZ:   true,
		SNZ:  true,
		SPL:  true,
		SMI:  true,
		MAX:  true,
		NMAX: true,
		TRUE: true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *CountNzCc) Condition() Condition {
	return this.condition
}
